File: prisma/seed/seed.ts
import { PrismaClient } from '@prisma/client';
import { techStacksData } from './techstacks-data';
import { categoriesData } from './categories-data';
import { domainsData } from './domains-data';
import { testUsersData } from './users-data';

const prisma = new PrismaClient();

async function seed() {
  console.log('Seeding tech stacks...');

  for (const techStack of techStacksData) {
    await prisma.techStack.upsert({
      where: { name: techStack.name },
      update: {
        iconUrl: techStack.iconUrl,
        type: techStack.type,
      },
      create: {
        name: techStack.name,
        iconUrl: techStack.iconUrl,
        type: techStack.type,
      },
    });
  }
  console.log(`✅ Seeded ${techStacksData.length} tech stacks`);

  console.log('Seeding categories...');
  for (const category of categoriesData) {
    await prisma.category.upsert({
      where: { name: category.name },
      update: {},
      create: {
        name: category.name,
      },
    });
  }
  console.log(`✅ Seeded ${categoriesData.length} categories`);

  console.log('Seeding domains...');
  for (const domain of domainsData) {
    await prisma.domain.upsert({
      where: { name: domain.name },
      update: {},
      create: {
        name: domain.name,
      },
    });
  }
  console.log(`✅ Seeded ${domainsData.length} domains`);

  // --- Test Users ---
  console.log('Seeding test users...');

  // Build lookup maps: name -> id
  const allTechStacks = await prisma.techStack.findMany();
  const tsMap = new Map(allTechStacks.map((t) => [t.name, t.id]));

  const allCategories = await prisma.category.findMany();
  const catMap = new Map(allCategories.map((c) => [c.name, c.id]));

  const allDomains = await prisma.domain.findMany();
  const domMap = new Map(allDomains.map((d) => [d.name, d.id]));

  for (const userData of testUsersData) {
    const user = await prisma.user.upsert({
      where: { email: userData.email },
      update: {
        name: userData.name,
        bio: userData.bio,
        jobTitle: userData.jobTitle,
      },
      create: {
        name: userData.name,
        email: userData.email,
        bio: userData.bio,
        jobTitle: userData.jobTitle,
      },
    });

    // Link tech stacks
    for (const tsName of userData.techStacks) {
      const tsId = tsMap.get(tsName);
      if (!tsId) {
        console.warn(`  ⚠ Tech stack "${tsName}" not found, skipping`);
        continue;
      }
      await prisma.userTechStack.upsert({
        where: { userId_techStackId: { userId: user.id, techStackId: tsId } },
        update: {},
        create: { userId: user.id, techStackId: tsId },
      });
    }

    // Link categories
    for (const catName of userData.categories) {
      const catId = catMap.get(catName);
      if (!catId) {
        console.warn(`  ⚠ Category "${catName}" not found, skipping`);
        continue;
      }
      await prisma.userCategories.upsert({
        where: { userId_categoryId: { userId: user.id, categoryId: catId } },
        update: {},
        create: { userId: user.id, categoryId: catId },
      });
    }

    // Link domains
    for (const domName of userData.domains) {
      const domId = domMap.get(domName);
      if (!domId) {
        console.warn(`  ⚠ Domain "${domName}" not found, skipping`);
        continue;
      }
      await prisma.userDomains.upsert({
        where: { userId_domainId: { userId: user.id, domainId: domId } },
        update: {},
        create: { userId: user.id, domainId: domId },
      });
    }

    console.log(`  ✅ ${userData.name} (${userData.techStacks.length} techs, ${userData.categories.length} cats, ${userData.domains.length} doms)`);
  }

  console.log(`✅ Seeded ${testUsersData.length} test users`);
}

async function main() {
  await seed();
}

main()
  .then(async () => {
    await prisma.$disconnect();
  })
  .catch(async (e) => {
    console.error(e);
    await prisma.$disconnect();
    process.exit(1);
  });

File: prisma/seed/seed.py
from prisma import Prisma
from pathlib import Path
import json
import sys

client = Prisma()

def find_model(client_obj, candidates):
    for name in candidates:
        if hasattr(client_obj, name):
            return getattr(client_obj, name)
    return None


def main():
    p = Path(__file__).with_name("techstacks-data.json")
    if not p.exists():
        print(f"Data file not found: {p}")
        sys.exit(1)

    data = json.loads(p.read_text())

    # connect the client (sync interface expected, generator uses interface = "sync")
    client.connect()
    try:
        # Try likely model attribute names
        model = find_model(client, ["tech_stack", "TechStack", "techstack"])
        if model is None:
            print("Could not find TechStack model on the Prisma client. Did you run `npx prisma generate`?")
            sys.exit(1)

        print("Seeding tech stacks...")
        for t in data:
            where = {"name": t["name"]}
            update = {"iconUrl": t["iconUrl"], "type": t["type"]}
            create = {"name": t["name"], "iconUrl": t["iconUrl"], "type": t["type"]}

            # Upsert using the model
            model.upsert(where=where, data={"create": create, "update": update})

        print(f"✅ Seeded {len(data)} tech stacks")

        # Seed Categories
        p_cat = Path(__file__).with_name("categories-data.json")
        if p_cat.exists():
            data_cat = json.loads(p_cat.read_text())
            model_cat = find_model(client, ["category", "Category"])
            
            if model_cat:
                print("Seeding categories...")
                for c in data_cat:
                    where = {"name": c["name"]}
                    create = {"name": c["name"]}
                    model_cat.upsert(where=where, data={"create": create, "update": {}})
                print(f"✅ Seeded {len(data_cat)} categories")
            else:
                print("⚠️  Category model not found, skipping categories.")
        else:
            print(f"⚠️  Categories data file not found: {p_cat}")
    finally:
        client.disconnect()


if __name__ == "__main__":
    main()
